rfc/halyard-lite.md

```diff
@@ -0,0 +1,475 @@
+Spinnaker using Halyard is to SSH to a VM (or a pod) and run a series of imperative
+commands.
+
+The desire for a Kubernetes-native path to install Spinnaker has led to a number
+of alternate install paths built on top of Halyard:
+*   A [Helm chart](https://github.com/helm/charts/tree/master/stable/spinnaker) that
+installs Halyard as a stateful set
+    *   While this has provided a Kubernetes-native way of installing Halyard,
+    the process of running and maintaining Spinnaker itself still involves SSH’ing
+    to the Halyard pod and running imperative commands
+*   An operator built by [OpsMx](https://operatorhub.io/operator/opsmx-spinnaker-operator)
+    *   This operator appears to have many of the same drawbacks as the Helm chart,
+    in that it will initially configure a Spinnaker installation but many customizations
+    still involve SSH’ing to a Halyard pod and running imperative commands.
+*   An operator built by [Armory](https://docs.armory.io/spinnaker/operator/)
+    *   This operator removes the need for users to run imperative commands to
+    manage Spinnaker.  It allows users to configure Spinnaker via CRUD on a
+    SpinnakerService CRD that contains the full Halyard configuration, with the
+    operator handling any redeployments from those changes.
+    *   While much of the surface area of Halyard is hidden from users of this
+    operator, it is still built on top of Halyard.
+
+### Duplication of Kubernetes-native tools
+
+Users often have requirements for customizing the Kubernetes objects that comprise
+a Spinnaker installation.  For example, a common use case is to add a toleration
+to some/all of the workloads so that they can run on specific node pools.
+
+As Halyard is responsible for both generating and deploying the YAML for all the
+Kubernetes objects that comprise a Spinnaker installation, Halyard needs to handle
+any customization that users want to apply.  Maintaining all of these possible
+customizations in Halyard adds a maintenance burden, as users frequently open pull
+requests to add additional fields they would like on their deployments/services.
+Changes to the templates that generate the YAML Halyard deploys are among the most
+frequent causes of bugs, as there is no test coverage on these files and the templating
+language (Jinja) is often not familiar to contributors.
+
+The problem of taking a set of YAML files that constitute a software installation,
+and customizing them with various overrides is one that has numerous solutions in
+the Kubernetes community, the most notable being [Helm](https://helm.sh/) and
+[Kustomize](https://github.com/kubernetes-sigs/kustomize).  Ideally, Halyard would
+not duplicate functionality that has existing solutions with much higher adoption;
+this would reduce both the maintenance burden on Halyard and the barrier to entry
+to Spinnaker for users familiar with these tools.
+
+### Dynamic configuration
```

Comment:
I think in general the approach is that config files can live in a source repository, and the running service can directly reference that repository.   In general, that works fine, as the config can be directly edited in the desired repo then be picked up by the service.

Where this doesn't interact well with Halyard is that it generates the service configs as transient files and then directly applies them; the generation of config is coupled to the deployment of that config.  After the changes in this RFC, I could imagine someone wanting to use Spring Cloud config deciding to commit the files generated by Halyard Lite to a config repo.  So the workflow would be:
* Edit the checked in hal config
* As part of the build step, run Halyard Lite to generate service configs, and also check those in
* Commit the full changes

Then when you deploy services, instead of deploying configs alongside them, just point them at your repo.

Let me know if this answers the question!